src/sqswatcher/sqswatcher.py

```diff
@@ -333,7 +336,10 @@ def _process_sqs_messages(
 
     for event in failed_events:
         log.error("Failed when processing event %s for instance %s", event.action, event.host)
-        _requeue_message(queue, event.message)
+        if event.action == "REMOVE":
+            _requeue_message(queue, event.message)
```

Comment:
The changes in _parse_sqs_messages ensure that during any one iteration of the sqswatcher loop, there won't be a REMOVE message processed after an ADD message for the same hostname. But does that cover stand across loop iterations as well? Is it possible that due to out of order delivery of SQS messages, we process an ADD during one iteration of the loop and then a REMOVE during the next iteration that was for a different instance ID?

Reply:
This is a good point. We are using SQS queue in standard mode meaning there is no guarantee that the order of messages is preserved: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/standard-queues.html#standard-queues-message-order

However the fact that an instance terminate for a given instance usually happens some time after a compute ready makes the reordering very unlikely to happen. In any case switching to a FIFO queue is probably the more correct thing to do. I would just consider this enhancement as out of the scope of this change if you agree.

In any case as long as ADD events are not retried and REMOVE events corresponding to given ADDs are processed after the ADD, the configuration of the scheduler should be kept consistent. Worst case scenario (that is very unlikely) we discard an ADD event for a legitimate node and we need to wait for the nodewatcher to replace it.